bundles/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/handler/EtherRainHandler.java

```diff
@@ -0,0 +1,213 @@
+ */
+@NonNullByDefault
+public class EtherRainHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(EtherRainHandler.class);
+
+    private @Nullable EtherRainCommunication device = null;
+    private boolean connected = false;
+    private @NonNullByDefault({}) EtherRainConfiguration config = null;
+
+    private @Nullable ScheduledFuture<?> updateJob = null;
+
+    private final HttpClient httpClient;
+
+    /*
+     * Constructor class. Only call the parent constructor
+     */
+    public EtherRainHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        this.httpClient = httpClient;
+        this.updateJob = null;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command == RefreshType.REFRESH) {
+            scheduler.execute(this::updateBridge);
+            return;
+        } else if (channelUID.getId().equals(EtherRainBindingConstants.CHANNEL_ID_EXECUTE)) {
+            execute();
+            updateState(EtherRainBindingConstants.CHANNEL_ID_EXECUTE, OnOffType.OFF);
+        } else if (channelUID.getId().equals(EtherRainBindingConstants.CHANNEL_ID_CLEAR)) {
+            clear();
+            updateState(EtherRainBindingConstants.CHANNEL_ID_CLEAR, OnOffType.OFF);
+        }
+    }
+
+    private boolean connectBridge() {
+        logger.debug("Attempting to connect to Etherrain with config = (Host: {}, Port: {}, Refresh: {}).", config.host,
+                config.port, config.refresh);
+
+        device = new EtherRainCommunication(config.host, config.port, config.password, httpClient);
+
+        try {
+            device.commandStatus();
+        } catch (EtherRainException | IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    "Could not create a connection to the EtherRain");
+            logger.debug("Could not open API connection to the EtherRain device. Exception received: {}",
+                    e.getMessage());
+            device = null;
+            updateStatus(ThingStatus.OFFLINE);
+            return false;
+        }
+
+        updateStatus(ThingStatus.ONLINE);
+
+        return true;
+    }
+
+    private void startUpdateJob() {
+        logger.debug("Starting Etherrain Update Job");
+        this.updateJob = scheduler.scheduleWithFixedDelay(this::updateBridge, 0, config.refresh, TimeUnit.SECONDS);
+
+        logger.debug("EtherRain sucessfully initialized. Starting status poll at: {}", config.refresh);
+    }
+
+    private void stopUpdateJob() {
+        logger.debug("Stopping Etherrain Update Job");
+
+        final ScheduledFuture<?> updateJob = this.updateJob;
+        if (updateJob != null && !updateJob.isDone()) {
+            updateJob.cancel(false);
+        }
+
+        this.updateJob = null;
+    }
+
+    @SuppressWarnings("null")
+    private boolean updateBridge() {
```

Comment:
I can foresee terrible things happening if this method was called concurrently by your scheduler and through a UI refresh command.

Reply:
Despite your initial comment, I didn't see anywhere else where you mentioned anything about synchronized on any other methods. However, not only did I add synchronized here, but everywhere else that sendGet was called as well as sendGet itself. Everything was tested and seem to work fine.  Let me know if there are more places where you think I should add it.

Reply:
Well, what I was trying to point out is that if it was called concurrently then you might end up calling `connectBridge()` concurrently as well. Which might result in your etherrain device causing one of requests to fail (maybe). But another possibility is that during a concurrent call an error causes the device field to be set to null, causing NPE for the other thread which expects it to be non-null at the time.

Synchronizing the methods of the `EtherRainCommunication` class can certainly prevent two threads from fighting over the same `EtherRainCommunication` instance, but it won't stop mishaps from occurring if each thread sets and nulls the device field underneath each other's feet.

But honestly, since your code is so simple the concurrency risk is very low, so I think just fixing the null warnings per my other comments would be more than sufficient to mitigate any concurrency issues. So other synchronization efforts on your part are not required. Just keep concurrency in mind in the future. 🙂